controllers/control/k8ssandratask_controller.go

```diff
@@ -0,0 +1,375 @@
+	"github.com/k8ssandra/k8ssandra-operator/pkg/clientcache"
+	"github.com/k8ssandra/k8ssandra-operator/pkg/config"
+	"github.com/k8ssandra/k8ssandra-operator/pkg/labels"
+	"github.com/k8ssandra/k8ssandra-operator/pkg/utils"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/cluster"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+	"strings"
+	"time"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	api "github.com/k8ssandra/k8ssandra-operator/apis/control/v1alpha1"
+)
+
+const (
+	k8ssandraTaskFinalizer = "k8ssandratask.k8ssandra.io/finalizer"
+	defaultTTL             = time.Duration(86400) * time.Second
+)
+
+var (
+	noTTL int32 = 0
+)
+
+// K8ssandraTaskReconciler reconciles a K8ssandraTask object:
+//   - create a CassandraTask for each DC specified by the K8ssandraTask. Those will be picked up and executed by
+//     cass-operator.
+//   - watch the CassandraTasks (see SetupWithManager), and update the K8ssandraTask's status accordingly.
+type K8ssandraTaskReconciler struct {
+	*config.ReconcilerConfig
+	client.Client
+	Scheme      *runtime.Scheme
+	ClientCache *clientcache.ClientCache
+	Recorder    record.EventRecorder
+}
+
+//+kubebuilder:rbac:groups=control.k8ssandra.io,namespace="k8ssandra",resources=k8ssandratasks,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=control.k8ssandra.io,namespace="k8ssandra",resources=k8ssandratasks/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=control.k8ssandra.io,namespace="k8ssandra",resources=k8ssandratasks/finalizers,verbs=update
+
+//+kubebuilder:rbac:groups=control.k8ssandra.io,namespace="k8ssandra",resources=cassandratasks,verbs=get;list;watch;delete
+//+kubebuilder:rbac:groups=control.k8ssandra.io,namespace="k8ssandra",resources=cassandratasks/status,verbs=get
+
+func (r *K8ssandraTaskReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx).WithValues("K8ssandraTask", req.NamespacedName)
+
+	logger.Info("Fetching task")
+	kTask := &api.K8ssandraTask{}
+	if err := r.Get(ctx, req.NamespacedName, kTask); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	kcKey := kTask.GetClusterKey()
+	kc, kcExists, err := r.getCluster(ctx, kcKey)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Set owner reference so that the task is deleted when the K8ssandraCluster is deleted.
+	if kTask.OwnerReferences == nil && kcExists {
```

Comment:
No need to check for nil here. The SetControllerReference checks that the owner of this task is the owner (controller - there's a difference here) that it's supposed to be and if not, update it.

Reply:
Yeah but SetControllerReference doesn't tell you if it had a side-effect. Without the nil check, I'll enter the loop and update every time, and never make it past this block.

Reply:
Do as many changes as possible in a single call to the API server, don't do extra calls.

Reply:
The owner reference is not in the status though, so I don't think the `r.Status().Update(...)` call at the end of the method would catch it.

Reply:
But the r.Update in the AddFinalizer is. Aren't they usually done in the same pass - object that hasn't been seen before?

Reply:
I refactored to do a single update.